pkg/kubelet/active_deadline.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+	"k8s.io/kubernetes/pkg/util"
+)
+
+const (
+	reason  = "DeadlineExceeded"
+	message = "Pod was active on the node longer than the specified deadline"
+)
+
+// activeDeadlineManager knows how to enforce active deadlines on pods.
+type activeDeadlineManager struct {
+	// the clock to use for deadline enforcement
+	clock util.Clock
+	// the provider of pod status
+	podStatusProvider status.PodStatusProvider
+	// the recorder to dispatch events when we identify a pod has exceeded active deadline
+	recorder record.EventRecorder
+}
+
+// newActiveDeadlineManager returns an active deadline manager that can enforce pod active deadline
+func newActiveDeadlineManager(
+	podLifecycleTarget lifecycle.PodLifecycleTarget,
+	podStatusProvider status.PodStatusProvider,
+	recorder record.EventRecorder,
+	clock util.Clock,
+) (*activeDeadlineManager, error) {
+
+	// check for all required fields
+	if clock == nil || podStatusProvider == nil || recorder == nil {
+		return nil, fmt.Errorf("Required arguments must not be nil: %v, %v, %v", clock, podStatusProvider, recorder)
+	}
+
+	// create the manager
+	manager := &activeDeadlineManager{
+		clock:             clock,
+		podStatusProvider: podStatusProvider,
+		recorder:          recorder,
+	}
+
+	// register the manager to participate in the kubelet sync loop, sync pod ops.
+	podLifecycleTarget.AddPodSyncLoopHandler(manager)
```

Comment:
This pattern makes the dependency tree complex. Why not return the manager here and inject it to pod sync lifecycle inside of kubelet object?